src/App.tsx

```diff
@@ -1,26 +1,125 @@
-/* eslint-disable max-len */
-/* eslint-disable jsx-a11y/control-has-associated-label */
-import React from 'react';
+import React, { useCallback, useEffect, useRef, useState } from 'react';
 import { UserWarning } from './UserWarning';
-
-const USER_ID = 0;
+import { getTodos, USER_ID } from './api/todos';
+import { Todo } from './types/Todo';
+import { Header } from './components/Header';
+import { TodoList } from './components/TodoList';
+import { Footer } from './components/Footer';
+import { ErrorMessage } from './components/ErrorMessage';
 
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [fetchedTodos, setFetchedTodos] = useState<Todo[]>([]);
+  const [errorMessage, setErrorMessage] = useState('');
+  const [filterBy, setFilterBy] = useState<'all' | 'active' | 'completed'>(
+    'all',
+  );
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [completedTodoIds, setCompletedTodoIds] = useState<number[] | null>(
+    null,
+  );
+  const [isHiddenError, setIsHiddenError] = useState(true);
+
+  const timeoutRef = useRef<number | null>(null);
+
+  const removeErrorMessage = () => {
+    setIsHiddenError(true);
+    if (timeoutRef.current) {
+      clearTimeout(timeoutRef.current);
+      timeoutRef.current = null;
+    }
+
+    timeoutRef.current = window.setTimeout(() => {
+      setErrorMessage('');
+    }, 1000);
+  };
+
+  const handleDelete = (id: number) => {
+    setTodos(prevTodo => prevTodo.filter(todo => todo.id !== id));
+  };
+
+  const errorMessageHandler = useCallback((er: Error) => {
+    setErrorMessage(er.message ?? String(er));
+    setIsHiddenError(false);
+
+    if (timeoutRef.current) {
+      clearTimeout(timeoutRef.current);
+    }
+
+    timeoutRef.current = window.setTimeout(() => {
+      removeErrorMessage();
+    }, 3000);
+  }, []);
+  // <br />
+  // Unable to update a todo
+
+  useEffect(() => {
+    getTodos()
+      .then(serverTodos => {
+        setTodos(serverTodos);
+        setFetchedTodos(serverTodos);
+      })
+      .catch(errorMessageHandler);
+  }, []);
+
+  useEffect(() => {
+    let filteredTodos = todos;
+
+    if (filterBy !== 'all') {
+      filteredTodos = todos.filter(todo =>
+        filterBy === 'active' ? !todo.completed : todo.completed,
+      );
+    }
+
+    setFetchedTodos([...filteredTodos]);
```

Comment:
why fetchedTodos if it's filtered todos?

But you don't need this state at all. It depends on other state values, so re-render will happen anyway - create a common variable `const visibleTodos = useMemo(...` and use useMemo hook to recalculate it only when it's necessary